src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/NameOfTests.fs

```diff
@@ -0,0 +1,219 @@
+        let tupledFunction(x,y) = x * y
+        let b = nameof tupledFunction
+        Assert.AreEqual("tupledFunction",b)
+
+    [<Test>]
+    member this.``local unit function name`` () =
+        let myFunction() = 1
+        let b = nameof(myFunction)
+        Assert.AreEqual("myFunction",b)
+
+    [<Test>]
+    member this.``local function parameter name`` () =
+        let myFunction parameter1 = nameof parameter1
+
+        Assert.AreEqual("parameter1",myFunction "x")
+
+    [<Test>]
+    member this.``can get name from inside a local function (needs to be let rec)`` () =
+        let rec myLocalFunction x = 
+            let z = 2 * x
+            nameof myLocalFunction + " " + z.ToString()
+            
+        Assert.AreEqual("myLocalFunction 46",myLocalFunction 23)
+        Assert.AreEqual("myLocalFunction 50",myLocalFunction 25)
+
+    [<Test>]
+    member this.CanGetNameFromInsideAMember () =
+        let b = nameof(this.CanGetNameFromInsideAMember)
+        Assert.AreEqual("CanGetNameFromInsideAMember",b)
+
+    [<Test>]
+    member this.``member function name`` () =
+        let b = nameof(this.MemberMethod)
+        Assert.AreEqual("MemberMethod",b)
+
+    [<Test>]
+    member this.``member function which is defined below`` () =
+        let b = nameof(this.MemberMethodDefinedBelow)
+        Assert.AreEqual("MemberMethodDefinedBelow",b)
+
+    member this.MemberMethodDefinedBelow(x,y) = x * y
+
+    [<Test>]
+    member this.``static member function name`` () =
+        let b = nameof(BasicNameOfTests.StaticMethod)
+        Assert.AreEqual("StaticMethod",b)
+
+    [<Test>]
+    member this.``class member lookup`` () =
+        let b = nameof(localConstant)
+        Assert.AreEqual("localConstant",b)
+
+    [<Test>]
+    member this.``member property name`` () =
+        let b = nameof(this.MemberProperty)
+        Assert.AreEqual("MemberProperty",b)
+
+    [<Test>]
+    member this.``static property name`` () =
+        let b = nameof(BasicNameOfTests.StaticProperty)
+        Assert.AreEqual("StaticProperty",b)
+
+    member this.get_XYZ() = 1
+
+    [<Test>]
+    member this.``member method starting with get_`` () =
+        let b = nameof(this.get_XYZ)
+        Assert.AreEqual("get_XYZ",b)
+
+    static member get_SXYZ() = 1
+
+    [<Test>]
+    member this.``static method starting with get_`` () =
+        let b = nameof(BasicNameOfTests.get_SXYZ)
+        Assert.AreEqual("get_SXYZ",b)
+
+    [<Test>]
+    member this.``nameof local property with encapsulated name`` () =
+        let ``local property with encapsulated name and %.f`` = 0
+        let b = nameof(``local property with encapsulated name and %.f``)
+        Assert.AreEqual("local property with encapsulated name and %.f",b)
+
+[<TestFixture>]
+type MethodGroupTests() =
+    member this.MethodGroup() = ()    
+    member this.MethodGroup(i:int) = ()
+
+    [<Test>]
+    member this.``method group name lookup`` () =
+        let b = nameof(this.MethodGroup)
```

Comment:
Looking at the impementation I think this only works because the first-class uses of ``this.MethodGroup`` defaults to resolving to the one-argument overload.  I think a truly overloaded set of methods may fail.  Could you add this test case please?

thanks